### PR TITLE
Introduce map reroll

### PR DESCRIPTION
**Pull request by PlayerNoon**
### Brief description of the changes:
Implementing feature approved by community.
https://discord.com/channels/823696400797138974/823771211421974579/980397648756490270
Players are able to vote to generate new map only within 30 s from the start (game time / ticks_played).
It require approval from 75% of all voters to pass.
Map is generated or not at 30 s mark from the (game time / ticks_played).
Players are able to generate new map as many times as they want.
Admin can disable this in confing panel.

### Tested Changes:
- [x] I've tested the changes with online players.
- [ ] I've not tested the changes.

### CODE REVIEW

**Comment by developer-8** [comment]
Why did you hide the 30 second countdown?
How to counter trolls who will click New map at the last second and the rest of the players simply ignore the vote? or  nothing can be done about it ?

**Comment by PlayerNoon** [comment]
> Why did you hide the 30 second countdown? How to counter trolls who will click New map at the last second and the rest of the players simply ignore the vote? or nothing can be done about it ?

I tried to have this code simple without unnecessary things. Also it was not mentioned / discussed on discord.
IMO I don't think it is an issue because it is just 30 s.
If it it will be a real problem then we will have to make poll about it on discord. 
But if u really want, tell how to make that timer so i can implement it.

File: maps/biter_battles_v2/difficulty_vote.lua [hunk 1]
```diff
@@ -114,11 +114,47 @@ local function on_player_left_game(event)
...
+	local yes = 0
+	local no = 0
+	for _, players in pairs(global.reroll_map_voting) do
+		if players == 1 then 
+			yes = yes + 1
+		else 
+			no = no + 1
+		end
+	end
+	local result = math.floor( yes / ( yes + no ) )
```
Comment (developer-8) [inline maps/biter_battles_v2/difficulty_vote.lua:127]: `result`  is always 1 or 0

File: maps/biter_battles_v2/difficulty_vote.lua [hunk 2]
```diff
@@ -124,14 +124,14 @@ local function set_reroll_map_voting_status(player)
 			no = no + 1
 		end
 	end
-	local result = math.floor( yes / ( yes + no ) )
-	if result > 0.75 then
+	local result = math.floor( yes / ( yes + no ) * 100 )
+	if result > 75 then
```
Comment (developer-8) [inline maps/biter_battles_v2/difficulty_vote.lua:128]:
```suggestion
	if result >= 75 then
```

File: maps/biter_battles_v2/game_over.lua [hunk 3]
```diff
@@ -495,5 +495,30 @@ local function chat_with_everyone(event)
     game.forces[enemy].print(message, player.chat_color)
 end
 
+function Public.reroll_map()
+    game.print("Generated new map")
+    game.speed = 1
+    global.reroll_voting_status = false
```
Comment (developer-8) [inline maps/biter_battles_v2/game_over.lua:501]:
```suggestion
```
it is not necessary

File: maps/biter_battles_v2/game_over.lua [hunk 4]
```diff
@@ -495,5 +495,30 @@ local function chat_with_everyone(event)
     game.forces[enemy].print(message, player.chat_color)
 end
 
+function Public.reroll_map()
```
Comment (developer-8) [inline maps/biter_battles_v2/game_over.lua:498]: `Public.reroll_map() ` consists of code from function `Public.server_restart()`, this needs to be done somehow so that the code is not duplicated

Comment (developer-8) [reply maps/biter_battles_v2/game_over.lua:498]: It seems to me that duplication is not necessary in this place, right?

Comment (PlayerNoon) [reply maps/biter_battles_v2/game_over.lua:498]: you right.

File: maps/biter_battles_v2/gui.lua [hunk 5]
```diff
@@ -39,6 +39,24 @@ function Public.clear_copy_history(player)
...
+function create_reroll_button(player)
+	if player.gui.top["reroll_yes"] then 
+		player.gui.top["reroll_yes"].destroy()
+		player.gui.top["reroll_no"].destroy()
+	end
+	if game.ticks_played < global.reroll_time_limit then
+		local b_reroll_yes = player.gui.top.add { type = "sprite-button", caption = "New map", name = "reroll_yes" }
+		b_reroll_yes.style.font = "heading-2"
+		b_reroll_yes.style.font_color = {r = 0.1, g = 0.9, b = 0.0}
+		gui_style(b_reroll_yes, {width = 150, height = 38})
```
Comment (developer-8) [inline maps/biter_battles_v2/gui.lua:51]: can pass `font` and `font_color` into `gui_style`

File: maps/biter_battles_v2/difficulty_vote.lua [hunk 6]
```diff
@@ -114,11 +114,47 @@ local function on_player_left_game(event)
 	set_difficulty()
 end
 
+local function set_reroll_map_voting_status(player)
+	local yes = 0
+	local no = 0
+	for _, players in pairs(global.reroll_map_voting) do
+		if players == 1 then
```
Comment (developer-8) [inline maps/biter_battles_v2/difficulty_vote.lua:121]:
```suggestion
	for _, vote in pairs(global.reroll_map_voting) do
		if vote == 1 then 
```
`players` confusing name

File: maps/biter_battles_v2/gui.lua (hunk 5 above)
Comment (TheBigZet) [reply maps/biter_battles_v2/gui.lua]: `gui_style` should work with every parameter from LuaStyle, it's just a for loop

File: maps/biter_battles_v2/game_over.lua [hunk 7]
```diff
@@ -495,5 +475,29 @@ local function chat_with_everyone(event)
...
```
Comment (developer-8) [inline maps/biter_battles_v2/game_over.lua:478]: Now that `reroll_map` used in `server_restart`, it can be renamed to something more suitable. At your discretion

File: maps/biter_battles_v2/main.lua [hunk 8]
```diff
@@ -156,6 +156,9 @@ local function on_tick()
 	if tick % 60 == 0 then 
 		global.bb_threat["north_biters"] = global.bb_threat["north_biters"] + global.bb_threat_income["north_biters"]
 		global.bb_threat["south_biters"] = global.bb_threat["south_biters"] + global.bb_threat_income["south_biters"]
+		if global.reroll_map_voting_status and game.ticks_played >= global.reroll_time_limit then
```
[the remark this answers is not in the document]
Comment (PlayerNoon) [reply maps/biter_battles_v2/main.lua:159]: image
 yes. first match it start form 0 tick. from next match it start form tick 11.

File: maps/biter_battles_v2/game_over.lua (hunk 7 above)
Comment (PlayerNoon) [reply maps/biter_battles_v2/game_over.lua:478]: i couldn't find any suitable name. you choose any.

File: maps/biter_battles_v2/game_over.lua [hunk 9]
```diff
@@ -495,5 +475,29 @@ local function chat_with_everyone(event)
...
+    game.print("Generated new map")
```
[the remark this answers is not in the document]
Comment (developer-8) [reply maps/biter_battles_v2/game_over.lua:479]: yes

File: maps/biter_battles_v2/game_over.lua (hunk 7 above)
Comment (developer-8) [reply maps/biter_battles_v2/game_over.lua:478]: how about `generate_new_map`

File: maps/biter_battles_v2/gui.lua [hunk 10]
```diff
@@ -39,6 +39,20 @@ function Public.clear_copy_history(player)
 	end
 end
 
+function create_reroll_button(player)
```
Comment (developer-8) [inline maps/biter_battles_v2/gui.lua:42]:
```suggestion
local function create_reroll_button(player)
```

**Comment by developer-8** [comment]
It may make sense to move the code from `difficulty_vote.lua` to `gui.lua` so that the creation of the buttons and the handling of their clicks happen in the same file.

**Comment by PlayerNoon** [comment]
done.

File: .vs/ProjectSettings.json [hunk 11]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (developer-8) [inline .vs/ProjectSettings.json:1]: remove folder .vs from git

Comment (developer-8) [reply .vs/ProjectSettings.json:1]: add it to .gitignore probably easiest

Comment (PlayerNoon) [reply .vs/ProjectSettings.json:1]: how

Comment (developer-8) [reply .vs/ProjectSettings.json:1]: to
https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/blob/master/.gitignore
add
.vs/

Comment (PlayerNoon) [reply .vs/ProjectSettings.json:1]: is it ok?

Comment (developer-8) [reply .vs/ProjectSettings.json:1]: yes, but .vs/ already commited so need to delete it 
You can do it on github

Comment (PlayerNoon) [reply .vs/ProjectSettings.json]: ok?

Comment (developer-8) [reply .vs/ProjectSettings.json]: ok

File: maps/biter_battles_v2/gui.lua [hunk 12]
```diff
@@ -541,6 +576,21 @@ local function on_gui_click(event)
 	if name == "bb_view_players" then
 		global.bb_view_players[player.name] = true
 		Public.create_main_gui(player)
+	end	
+
+	if event.element.name == "reroll_yes" then
```
Comment (developer-8) [inline maps/biter_battles_v2/gui.lua:581]:
```suggestion
	if name == "reroll_yes" then 
```

File: maps/biter_battles_v2/gui.lua [hunk 13]
```diff
@@ -541,6 +576,21 @@ local function on_gui_click(event)
...
+	end	
+
+	if event.element.name == "reroll_yes" then 
+		if global.reroll_map_voting[player.name] ~= 1 then 
+			global.reroll_map_voting[player.name] = 1 
+			game.print(player.name .. " want to reroll map ",{r = 0.1, g = 0.9, b = 0.0})	
+			set_reroll_map_voting_status()
+		end
+	end
+	if event.element.name == "reroll_no" then
```
Comment (developer-8) [inline maps/biter_battles_v2/gui.lua:588]:
```suggestion
	if name == "reroll_no" then 
```

File: maps/biter_battles_v2/main.lua [hunk 14]
```diff
@@ -156,6 +156,10 @@ local function on_tick()
 	if tick % 60 == 0 then 
 		global.bb_threat["north_biters"] = global.bb_threat["north_biters"] + global.bb_threat_income["north_biters"]
 		global.bb_threat["south_biters"] = global.bb_threat["south_biters"] + global.bb_threat_income["south_biters"]
+		if global.reroll_map_voting_status and game.ticks_played >= global.reroll_time_limit then 
+		Game_over.generate_new_map()	
+		game.print("Generated new map")
```
Comment (developer-8) [inline maps/biter_battles_v2/main.lua:161]:
```suggestion
			Game_over.generate_new_map()	
			game.print("Generated new map")
```

File: maps/biter_battles_v2/gui.lua [hunk 15]
```diff
@@ -470,6 +485,26 @@ function Public.spy_fish()
...
+		end
+	end
+	local result = math.floor( yes / ( yes + no ) * 100 )
+	if result >= 75 then
+		global.reroll_map_voting_status = true
+	else
+		global.reroll_map_voting_status = false
+	end
+	local reroll_time_left = math.floor((global.reroll_time_limit - game.ticks_played)/60)
+	game.print(result .. "% votes to reroll." .. " Need 75% to reroll map.".. " Time left " .. reroll_time_left .. "s" )
```
Comment (developer-8) [inline maps/biter_battles_v2/gui.lua:505]:
```suggestion
	game.print(result .. "% votes to reroll. Need 75% to reroll map. Time left " .. reroll_time_left .. "s" )
```

**Review by developer-8 — APPROVED** [review]

**Comment by TheBigZet** [comment]
I'd like having a toggle in config in case the system is abused

File: maps/biter_battles_v2/init.lua [hunk 16]
```diff
@@ -112,6 +112,7 @@ function Public.initial_setup()
 	}
 	for _, d in pairs(defs) do p.set_allows_action(d, true) end
 
+	global.reroll_time_limit = 1800
```
Comment (TheBigZet) [inline maps/biter_battles_v2/init.lua:115]: I think it's too short, I get loaded around 800th tick, leaving about 15s to check the map and vote

**Review by TheBigZet — CHANGES_REQUESTED** [review]
I'd like to try to do it with `utils/Task.lua`
Set function checking `reroll_voting_status` to be run x seconds after map start. This way we can cut down number of checks per tick.

I haven't used that module yet tho, not sure if it works like that

File: maps/biter_battles_v2/init.lua [hunk 17]
```diff
@@ -121,6 +122,7 @@ function Public.initial_setup()
 		["only_admins_vote"] = false,		--Are only admins able to vote on the global difficulty?
 		--MAP SETTINGS--
 		["new_year_island"] = false,
+		["map_reroll_admin_disable"] = true,
```
[the remark this answers is not in the document]
Comment (Ragnarok77-factorio) [reply maps/biter_battles_v2/init.lua:125]: Very minor but agree too, would prefer enabled too, less confusing

**Comment by TheBigZet** [comment]
Button creation and removal could also be done just once, as well as ratio calculation.

**Comment by developer-8** [comment]
I got a message about [0c7c584b2d632ec0f929034729cfdfb6356b2e83](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/commit/0c7c584b2d632ec0f929034729cfdfb6356b2e83) but I don't see it in the commits list

**Comment by TheBigZet** [comment]
> I got a message about [0c7c584b2d632ec0f929034729cfdfb6356b2e83](https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/commit/0c7c584b2d632ec0f929034729cfdfb6356b2e83) but I don't see it in the commits list

Yeah weird... It shows up in the source branch PlayerNoon:reroll-previev

**Review by developer-8 — COMMENTED** [review]
1. Map will not reroll if this is first game and `generate_new_map()` was never called before.

2. Map will not reroll if click "Mew map"  after time out but GUI is not hidden yet.

      before Task changes:
      image
      after Task map not rerolled:
      image

**Comment by Ragnarok77-factorio** [comment]
Waiting for fix of developer comment before testing it again

**Comment by developer-8** [comment]
I'm not insist for it to be fixed. It works well enough, no complaints about the code, bugs are minor. There's just room for improvement. I leave it to the discretion of the author

**Review by Ragnarok77-factorio — APPROVED** [review]
Tested it, fine for me now.
But it was a solo test.

I think it would be nice to test it a bit on pub server just to be sure no edge case was forgotten or something.

**Comment by Ragnarok77-factorio** [comment]
We rerolled it, but it's quite spammy too be honest, and may the timing is a bit too short too, barely any time to vote..? Dunno, just a random opinion, what do you think? 

image

**Comment by Ragnarok77-factorio** [comment]
For the trace, it was tested live on bb, no issue found it seems like.

**Review by Ragnarok77-factorio — APPROVED** [review]
Tested for months on bb main server , no issue, ok for the rest too.
